src/providers/wfs/qgsbackgroundcachedfeatureiterator.cpp

```diff
@@ -0,0 +1,870 @@
+  for ( const QgsFeatureUniqueIdPair &pair : constList )
+  {
+    *mWriterStream << pair.first;
+  }
+  if ( !mWriterFile && mWriterByteArray.size() > mWriteTransferThreshold )
+  {
+    QString thisStr;
+    thisStr.sprintf( "%p", this );
+    ++ mCounter;
+    mWriterFilename = QDir( mShared->acquireCacheDirectory() ).filePath( QStringLiteral( "iterator_%1_%2.bin" ).arg( thisStr ).arg( mCounter ) );
+    QgsDebugMsgLevel( QStringLiteral( "Transferring feature iterator cache to %1" ).arg( mWriterFilename ), 4 );
+    mWriterFile = new QFile( mWriterFilename );
+    if ( !mWriterFile->open( QIODevice::WriteOnly | QIODevice::Truncate ) )
+    {
+      QgsDebugMsg( QStringLiteral( "Cannot open %1 for writing" ).arg( mWriterFilename ) );
+      delete mWriterFile;
+      mWriterFile = nullptr;
+      mWriterFilename.clear();
+      mShared->releaseCacheDirectory();
+      return;
+    }
+    mWriterFile->write( mWriterByteArray );
+    mWriterByteArray.clear();
+    mWriterStream->setDevice( mWriterFile );
+  }
+}
+
+// This will invoked asynchronously, in the thread of QgsBackgroundCachedFeatureIterator
+// hence it is safe to quite the loop
+void QgsBackgroundCachedFeatureIterator::featureReceived( int /*featureCount*/ )
+{
+  //QgsDebugMsg( QStringLiteral("QgsBackgroundCachedFeatureIterator::featureReceived %1 features").arg(featureCount) );
+  if ( mLoop )
+    mLoop->quit();
+}
+
+void QgsBackgroundCachedFeatureIterator::checkInterruption()
+{
+  //QgsDebugMsg("QgsBackgroundCachedFeatureIterator::checkInterruption()");
+
+  if ( mInterruptionChecker && mInterruptionChecker->isCanceled() )
+  {
+    mDownloadFinished = true;
+    if ( mLoop )
+      mLoop->quit();
+  }
+}
+
+void QgsBackgroundCachedFeatureIterator::timeout()
+{
+  mTimeoutOccurred = true;
+  mDownloadFinished = true;
+  if ( mLoop )
+    mLoop->quit();
+}
+
+bool QgsBackgroundCachedFeatureIterator::fetchFeature( QgsFeature &f )
+{
+  f.setValid( false );
+
+  if ( mClosed )
+    return false;
+
+  // First step is to iterate over the on-disk cache
+  QgsFeature cachedFeature;
+  while ( mCacheIterator.nextFeature( cachedFeature ) )
+  {
+    if ( mInterruptionChecker && mInterruptionChecker->isCanceled() )
+      return false;
+
+    if ( mTimeoutOccurred )
+      return false;
+
+    //QgsDebugMsg(QString("QgsBackgroundCachedSharedData::fetchFeature() : mCacheIterator.nextFeature(cachedFeature)") );
+
+    if ( mShared->hasGeometry() && mFetchGeometry )
+    {
+      int idx = cachedFeature.fields().indexFromName( QgsBackgroundCachedFeatureIteratorConstants::FIELD_HEXWKB_GEOM );
+      Q_ASSERT( idx >= 0 );
+
+      const QVariant &v = cachedFeature.attributes().value( idx );
+      if ( !v.isNull() && v.type() == QVariant::String )
+      {
+        QByteArray wkbGeom( QByteArray::fromHex( v.toString().toLatin1() ) );
+        QgsGeometry g;
+        unsigned char *wkbClone = new unsigned char[wkbGeom.size()];
+        memcpy( wkbClone, wkbGeom.data(), wkbGeom.size() );
```

Comment:
I don't think this copy is required -- there's a QgsGeometry::fromWkb which takes a QByteArray value already (and the char* version doesn't take ownership in any case)

Reply:
indeed. fromWkb( const QByteArray &wkb ) was added in 3.0. This code predates it :-) . Taken into account.
(note you mention that the char* version doesn't take ownership, but the code says the contrary)

Reply:
Argh, you're right. Too early in the day for PR reviews :P